[YUNIKORN-2325] Add a chart to display multi-type resource utilisation (Web)

chenyulin0719 opened the pull request:
### What is this PR for?
Create a vertical bar chart to display multi-type resource utilisation in YuniKorn UI Nodes page.
- Add 2 componenets,  app-node-utilizations and vertical-bar-chart
- Calculate overall resource usage for each resource type, put top 10 resources in legend  (Order by avg resource usage)
- Only display domaint type's node utilization by default, could display others through legend.


### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
- Remove domaint reosurce chart in dashboard page.

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-2325

### How should this be tested?
- make lint
- make license-check
- yarn test:coverage
- yarn start:srv
- yarn start   (Check the Nodes page)

### Screenshots (if appropriate)
![YUNIKORN-2325 Demo](https://github.com/apache/yunikorn-web/assets/26764036/fbf5486c-0bc0-4c96-934d-16ce9bc4018e)

Recording: https://www.loom.com/share/657bde1a7d834c29b943fbcf70a3ee9a?sid=e6892e8f-c258-4f04-8d3f-1c72435f08c4

### Questions:
NA

craigcondit reviewed (changes requested):
Some minor nits, and a couple of thoughts:

The bar chart is quite large, but the legend at the top feels squished. Perhaps instead of arranging it in a single row with {color, item}, {color, item}..., it might make sense to put one item per row and arrange the legend vertically. This would also allow adding a checkmark (for selected) or an X (for unselected) next to each entry to indicate visible state rather than a strikethrough text effect. I think this will look cleaner. We can decrease the overall height of the bar chart to compensate (and even more -- ideally it would still fit on one page on a typical browser window).

Minor nit on the vertical hashmarks - the values here are always integers since the values are a count of nodes in each bucket. So, we can hid the decimal (show 1 instead of 1.0, etc.)

Otherwise, this looks really good.

chia7712 commented:
<img width="810" alt="截圖 2024-01-28 凌晨12 19 32" src="https://github.com/apache/yunikorn-web/assets/6234750/c9eabbbc-dc18-43ad-bbd0-4ba98914a1ee">

`Nodes Resource Utilization` in the dashboard page gets "undefined"

chia7712 commented:
<img width="1172" alt="截圖 2024-01-28 凌晨12 21 23" src="https://github.com/apache/yunikorn-web/assets/6234750/e7d77ca3-4183-4554-8d22-57adacc1530c">

the word "Domaint Resource: memory" is weird since we can select multi-resources now.

craigcondit commented:
> <img width="1172" alt="截圖 2024-01-28 凌晨12 21 23" src="https://github.com/apache/yunikorn-web/assets/6234750/e7d77ca3-4183-4554-8d22-57adacc1530c">
> 
> 
> 
> the word "Domaint Resource: memory" is weird since we can select multi-resources now.

Agreed. I think we can remove it. It's also mis-spelled (should be Dominant).

chenyulin0719 commented:
Hi @craigcondit, @chia7712 ,
Chart.js doesn't support a checkbox legend, so I customized a div block and added checkboxes in.

Below are the new changes:
1. Implement customize legend with checkbox.
2. Rename title to ‘Node Resource Utilization’
3. Change the y-axis nodes scales to intetger
4. Change the heigh of barchart canvas to 50vh
5. Remove “Dominant Resources:xxx” sub-title, but add a text before checkbox to emphasize “Resources are sorted by load”. (Please let me know if you think it’s too verbose)

@chia7712 
> The "Nodes Resource Utilization" in the dashboard page gets "undefined"

This is an issue of the old API: ws/v1/scheduler/node-utilization .The API returns empty resources if the cluster has never run any workload before.
The API was mark as deprecated in [YUNIKORN-2257](https://issues.apache.org/jira/browse/YUNIKORN-2257 ) , and I planned to remove the dashboard chart in [[YUNIKORN-2339] Remove Nodes Utilisation chart from Dashboard page (Web)](https://issues.apache.org/jira/browse/YUNIKORN-2339)


![YUNIKORN-2325 demo2](https://github.com/apache/yunikorn-web/assets/26764036/48ce1a51-1d12-415e-b9d8-98a3549a5de9)

chenyulin0719 commented on src/app/components/app-node-utilizations/app-node-utilizations.component.html:
```diff
@@ -0,0 +1,31 @@
+
+<mat-card appearance="outlined" class="box-card">
+  <mat-card-header>
+    <mat-card-title>Node Utilizations</mat-card-title>
```
Fixed.

chenyulin0719 commented on src/app/components/app-node-utilizations/app-node-utilizations.component.html:
```diff
@@ -0,0 +1,31 @@
+<mat-card appearance="outlined" class="box-card">
+  <mat-card-header>
+    <mat-card-title>Node Utilizations</mat-card-title>
+    <mat-card-subtitle>Domaint Resource: {{domaintResource}}</mat-card-subtitle>
```
Removed the "Domaint Resource" subtitle.

chia7712 commented:
> but add a text before checkbox to emphasize “Resources are sorted by load”. (Please let me know if you think it’s too verbose)

pardon me, I failed to catch the point of "high to low"...

chia7712 commented:
<img width="318" alt="截圖 2024-01-30 晚上7 34 44" src="https://github.com/apache/yunikorn-web/assets/6234750/667b285c-1e45-403a-a78d-bf17050843d2">

another idea: how about displaying all items by default? currently only `vcore` is selected.

chia7712 commented:
The color for each item gets changed after the page is refresh. Is it expected? Also, is it more readable if we show the relation between resource item and color?

chenyulin0719 commented:
HI @chia7712 ,
> pardon me, I failed to catch the point of "high to low"...

The checkboxes are displayed in the order of avg resource loading.  
The first resource is the "dominant resource" (The resource type which have the highest utilization.). 

> another idea: how about displaying all items by default? currently only vcore is selected.

Current I only display the "Dominant resource" by default. I believe displaying other resources would be too cluttered.
Showing the dominant resource helps users know what resource is the most loaded.
Let user make decision for displaying other resources.

> The color for each item gets changed after the page is refresh. Is it expected? Also, is it more readable if we show the relation between resource item and color?

The color is base on the  items order of /ws/v1/scheduler/node-utilizations.  I didn't sorted it in core because I thought there is no obvious order for the resource types.
But I agree that adding a fixed color would be better. I think sorting by resource type name (by lexicographical) could be a reasonable sorting.

For exampe, if I have 3 resource (GPU, Memory, CPU) in YuniKorn scheduler. The color mapping of each resource will be:
1. CPU -> CHART_COLORS[0]
2. Memory -> CHART_COLORS[1]
3. GPU -> CHART_COLORS[2]

@chia7712 Let me know if you have other ideas for the sorting. If not I will change add a commit to sort it.

chenyulin0719 commented:
Hi @chia7712,
Just update this PR. Set fixed color based on lexicographically order of resource types

![YUNIKORN-2325 Demo 3](https://github.com/apache/yunikorn-web/assets/26764036/2d26c279-3d58-4570-9dbe-fe188efbcb69)

wusamzong commented on src/app/components/app-node-utilizations/app-node-utilizations.component.ts:
```diff
@@ -0,0 +1,150 @@
+    // sort by resource type first, then sort by avg utilization rate
+    barChartDataSets.sort((a, b) => CommonUtil.resourcesCompareFn(a.label, b.label));
+    barChartDataSets.sort((a, b) => b.avgUtilizationRate - a.avgUtilizationRate);
+    barChartDataSets = barChartDataSets.slice(0, 5); // only show top 5 resources
```
Since the legend uses checkboxes, freeing up extra space for resource types, is it still necessary to limit the display to only the top 5 resources?

chenyulin0719 replied:
Hi @wusamzong, Agree to show more resourced because we allow users to select what they want to see,  I think change the limit to 10 is acceptible in the current layout.

chenyulin0719 replied:
Changed the maximum resource type number to 10.

(6 types in my environement)
<img width="386" alt="image" src="https://github.com/apache/yunikorn-web/assets/26764036/5d6159fa-d054-4586-9d26-cb06d0998c4d">

doupache commented on src/app/components/vertical-bar-chart/vertical-bar-chart.component.ts:
```diff
@@ -0,0 +1,198 @@
+        chartLegendDiv.appendChild(document.createTextNode('Resource sorted by load:'));
+        chartLegendDiv.appendChild(document.createElement('br'));
+        chartLegendDiv.appendChild(document.createTextNode('(high to low)'));
+        chartLegendDiv.appendChild(document.createElement('br'));
```
nit: Resource sorted by load: -> Top 10 resources sorted by load (high to low)

chenyulin0719 replied:
Just updated the description.  Thanks for the feedback.

doupache reviewed (commented):
Great work! 🎉 
Just a small nitpick:  considering that we exclusively display the top 10 resources, we might want to update the legend description to read 'Top 10 resources sorted by load (hight to low)' 

Other than that, it looks good to me!

chenyulin0719 commented:
Hi @doupache , @chia7712 ,
Just added color box and the description. Thanks for the feedbacks.

<img width="1137" alt="image" src="https://github.com/apache/yunikorn-web/assets/26764036/8e389669-450b-40fc-840e-149e8aa04cfe">

craigcondit commented:
I wonder if we can simplify a bit here. Do we even need to deselect resources? I think there's plenty of room for all of them. We can also make the description simpler (Top 10 resources).

chenyulin0719 commented:
Hi @craigcondit, 
I believe having an option to deselect resources is better to have because if a user just want to check the distriibution of memory's node utilization,  deselecting other resources could make the chart cleaner. 
(We can change the default selected types number, I set it to top 1 becasue I assumed that checking dominant reosurec is the most common use case of this chart)

Regarding the description:
- Top 10 resources 
- Top 10 resources **sorted by load**:
While the addition of "sorted by load" might seem verbose, but I think it's necessarily to convey the idea that the resources are ranked based on their loading in nodes. Otherwise, users won't know that the first item represents the dominant resource.

craigcondit commented:
> Hi @craigcondit, I believe having an option to deselect resources is better to have because if a user just want to check the distriibution of memory's node utilization, deselecting other resources could make the chart cleaner. (We can change the default selected types number, I set it to top 1 becasue I assumed that checking dominant reosurec is the most common use case of this chart)

Less is more. We have only a maximum of 10 resources (typical will be less) and 10 buckets (0-10%, 10-20%, etc). The color coding will make this visually easy to distinguish without adding extra logic. If you're concerned with being able to distinguish between the bars, add a mouse hover on each bar that changes the color to something slightly brighter for all bars of the same type. 

> 
> Regarding the description:
> 
> * Top 10 resources
> * Top 10 resources **sorted by load**:
>   While the addition of "sorted by load" might seem verbose, but I think it's necessarily to convey the idea that the resources are ranked based on their loading in nodes. Otherwise, users won't know that the first item represents the dominant resource.

Again, less is more. The sort order will be apparent by the ordering of the bars. We're making this more complex than it needs to be. Graphs should be self-explanatory; in fact, I really don't even think we need a descriptor there at all. The "things" we are displaying are obviously resources. The fact that it's a top-10 list will not even matter 99% of the time.

chia7712 commented:
> add a mouse hover on each bar that changes the color to something slightly brighter for all bars of the same type.

that is good idea to me. We can change the opaqueness of other elements to 0

> Again, less is more. The sort order will be apparent by the ordering of the bars. We're making this more complex than it needs to be. Graphs should be self-explanatory; in fact, I really don't even think we need a descriptor there at all. The "things" we are displaying are obviously resources. The fact that it's a top-10 list will not even matter 99% of the time.

I'm +1 on this. The description is too verbose to understand :(

craigcondit commented:
> > add a mouse hover on each bar that changes the color to something slightly brighter for all bars of the same type.
> 
> that is good idea to me. We can change the opaqueness of other elements to 0
>
0 would be invisible. Perhaps 50%? That will emphasize the one that's being looked at but still keep the others partially visible.

chia7712 commented:
> 0 would be invisible. Perhaps 50%? That will emphasize the one that's being looked at but still keep the others partially visible.

+1

chenyulin0719 commented:
Hi @craigcondit , @chia7712 ,

Just update a newer version, below are the changes:
1. Remove the verbose description
2. Change bar colors of same type when a mouse hover. (opaqueness: 1 -> 0.5)
3. Change the default item size from 1 to 5. (Based on the [original discussion on Slack](https://yunikornworkspace.slack.com/archives/CL9CRJ1KM/p1701742319324709?thread_ts=1701680922.070739&cid=CL9CRJ1KM), I can change it if required.)

> Less is more. We have only a maximum of 10 resources (typical will be less) and 10 buckets (0-10%, 10-20%, etc). The color coding will make this visually easy to distinguish without adding extra logic. If you're concerned with being able to distinguish between the bars, add a mouse hover on each bar that changes the color to something slightly brighter for all bars of the same type.

I still kept the checkbox for below reasons:
1. In some use cases, users might only care about 2~3 resources (ex: CPU/Memory/Pod).  Even though we have the hover feature, it could only highlight a single resource. 
2. I set a default display item size to 5,  more resource will be just cluttering. But I think it's no harm to keep the flexibility to select/deselect more resoueces.

@chia7712, @craigcondit,
I can remove the checkbox if I got -1 for the above considerations. Otherwise I think we can keep it.
Thanks for your feedback!

![YUNIKORN-2325 Demo 4](https://github.com/apache/yunikorn-web/assets/26764036/c7820830-b8ba-4bc8-896e-accd9af64e4f)

craigcondit commented:
> @chia7712, @craigcondit, I can remove the checkbox if I got -1 for the above considerations. Otherwise I think we can keep it. Thanks for your feedback!

Let's remove the checkbox. Between the color differences and the hover behavior, I think the new design is perfectly clear and there's no need to  deselect items.

chenyulin0719 commented:
@craigcondit,
OK, I'll remove the checkbox. 
One more question, for the display item size, should I display all resources or just show top 5?

craigcondit commented:
> @craigcondit, OK, I'll remove the checkbox. One more question, for the display item size, should I display all resources or just show top 5?

I think top 5 (or 10) should be fine. Most clusters won't have more than 10, so that would show everything in that case. Just looking at an internal cluster, the default resources include pods, cpu, memory, ephemeral-storage, hugepages-1Gi and hugepages-2Mi. That's 6, but of those, typically only memory and cpu are used for quotas. Some clusters may have custom resources such as GPU as well, but there's unlikely to be > 10 that are ever in contention.

chenyulin0719 commented:
Hi @craigcondit ,
Just updated a newer version. 
- Remove check box div and changed back to chart js ledgend

I keep the display item size to 10.

<img width="1458" alt="image" src="https://github.com/apache/yunikorn-web/assets/26764036/e85a6ba3-7389-4697-8c65-ac79366c690f">

craigcondit reviewed (approved):
+1 LGTM.
